bpo-30555: Fix WindowsConsoleIO errors in the presence of fd redirection

segevfiner opened the pull request:
This works by not caching the handle and instead getting the handle from the file descriptor each time, so that if the actual handle changes by fd redirection closing/opening the console handle beneath our feet, we will keep working correctly.

I think I also fixed some resource handling issues along the way since ucrt/msvcrt fds *own* the HANDLEs they use and this module tried to close them itself.

See https://github.com/pytest-dev/py/issues/103 for the issue it causes in Pytest. Also see: https://github.com/pytest-dev/pytest/pull/2462. I was able to run the Pytest test suite, and my own little test described in the PR, to completion after this change.

Having your console handle closed by fd redirection is likely a change in the ucrt/msvcrt or a change in newer Windows versions (Console handles were rewritten in Windows 8).

This likely also affects any other module caching `GetStdHandle` or `_get_osfhandle`. For example colorama (https://github.com/pytest-dev/pytest/issues/2465 - https://github.com/tartley/colorama/pull/131), and I can think of "pyreadline", "win-unicode-console" and "click" probably suffering from the same issue but I haven't tested them.

I'm also not sure of the performance implications of this change.

This affects Python 3.6-3.7

P.S. `python -m test -R 3:2 test.test_winconsoleio` reports ref leaks on master for me 😛

CC @zooba @eryksun

- bpo-30555

segevfiner committed:
bpo-30555: Fix WindowsConsoleIO failure in the presence of fd redirection

This works by not caching the handle and instead getting the handle from
the file descriptor each time, so that if the actual handle changes by
fd redirection closing/opening the console handle beneath our feet, we
will keep working correctly.

I think I also fixed some resource handling issues along the way since
ucrt/msvcrt fds *own* the HANDLEs they use and this module tried to
close them itself.

segevfiner committed:
bpo-30555: Fixed the kludgey PC/_testconsole.c, grumble grumble...

eryksun commented:
Most of the reference leaks are due to initializing [`decodedname = Py_None`](https://github.com/segevfiner/cpython/blob/ac3af9ec5e4bf8c18243ab7c8bd6bacf2317a2b6/Modules/_io/winconsoleio.c#L311) followed by  `Py_INCREF(decodedname)`. It should be initialized as `NULL`. 

Some of the reported ref leaks are from calling `getwindowsversion` in [`test_conout_path`](https://github.com/segevfiner/cpython/blob/ac3af9ec5e4bf8c18243ab7c8bd6bacf2317a2b6/Lib/test/test_winconsoleio.py#L112). This call can be moved out of the test by using a global variable, or by splitting the test up like this:

    def conout_path_test(self, dos_device):
        temp_path = tempfile.mkdtemp()
        self.addCleanup(support.rmtree, temp_path)
        conout_path = os.path.join(temp_path, 'CONOUT$')

        with open(conout_path, 'wb', buffering=0) as f:
            if dos_device:
                self.assertIsInstance(f, ConIO)
            else:
                self.assertNotIsInstance(f, ConIO)

    @unittest.skipIf(sys.getwindowsversion()[:2] <= (6, 1),
        "Windows 7 and earlier resolve CONOUT$ in a directory as a disk file")
    def test_conout_path_device(self):
        self.conout_path_test(dos_device=True)

    @unittest.skipIf(sys.getwindowsversion()[:2] > (6, 1),
        "Windows 8 and later resolve CONOUT$ in a directory as a DOS device")
    def test_conout_path_file(self):
        self.conout_path_test(dos_device=False)

eryksun commented on Modules/_io/winconsoleio.c:
```diff
@@ -133,6 +133,21 @@ char _PyIO_get_console_type(PyObject *path_or_fd) {
     return m;
 }
 
+static HANDLE py_get_osfhandle(int fd)
```
`_Py_get_osfhandle`, `_Py_get_osfhandle_noraise`, `_Py_open_osfhandle`, and `_Py_open_osfhandle_noraise` could be added in Python/fileutils.c. These CRT functions are called in fileutils.c, msvcrtmodule.c, posixmodule.c, mmapmodule.c, myreadline.c, _testconsole.c, and here.

segevfiner replied:
Good idea.

Do you wish me to:
1. Do this entire refactor right here in this PR.
2. Add the functions and only use them in winconsoleio.c and _testconsole.c which I touched in this PR and leave the rest of the modules to a separate PR.
3. Leave this to a separate PR entirely.

?

segevfiner replied:
I implemented this, with the full replacement in a separate commit just in case it's frowned upon by a core developer during review.

segevfiner commented:
@eryksun So the leaks are real than, nice find! 😄 I can submit a PR/issue with your suggested fix if it helps you or just leave it to you. Why would calling `sys.getwindowsversion` in a test function leak though? Does that function leak? Or is there something else at play?

**EDIT:** I can also apply your fix right here if you wish me to.

eryksun commented:
It should be ok to fix the `_WindowsConsoleIO` leak and rewrite the test in this issue. The ref leak in `getwindowsversion` needs a new issue. If you want to work on it, the problem is that it calls  `PyTuple_Pack` directly on the `int` objects created from `realMajor`, etc.  `PyTuple_Pack` increments the ref count of each object instead of stealing it.

segevfiner committed:
bpo-30555: Fix refleaks

The change in test_winconsoleio.py is only to avoid the ref leak in
`sys.getwindowsversion`.

segevfiner committed:
bpo-30555: Add _Py_{get,open}_osfhandle{,_noraise} and use them instead

segevfiner committed:
bpo-30555: Swapped the rest of the uses of _{get,open}_osfhandle

eryksun commented on Include/fileutils.h:
```diff
@@ -117,7 +117,15 @@ PyAPI_FUNC(int) _Py_dup(int fd);
 PyAPI_FUNC(int) _Py_set_blocking(int fd, int blocking);
-#endif   /* !MS_WINDOWS */
+#else   /* MS_WINDOWS */
+PyAPI_FUNC(intptr_t) _Py_get_osfhandle_noraise(int fd);
```
I thought one of benefits would be using `HANDLE` in the signature instead of `intptr_t`, to avoid all of the casts. I guess we can't. How about `void *` like `_PyOS_SigintEvent`?

segevfiner replied:
I think CPython explicity doesn't include Windows.h anywhere in it's public headers. fileutils.h is included by Python.h so I don't think I can really include Windows.h there. Ae such I don't have HANDLE defined and assuming that someone included Windows.h by himself before this header breaks horribly of course.

eryksun replied:
[`HANDLE`](https://msdn.microsoft.com/en-us/library/aa383751#HANDLE) is a `typedef` for `void *`, so we can use `void *` instead, like `_PyOS_SigintEvent` does.

segevfiner committed:
bpo-30555: Swap intptr_t to void* in _Py_{get,open}_osfhandle{,_noraise}

eryksun commented on PC/msvcrtmodule.c:
```diff
@@ -167,15 +167,7 @@ static long
 msvcrt_open_osfhandle_impl(PyObject *module, intptr_t handle, int flags)
```
The clinic definition could be updated here, e.g. to use a `HANDLE` type based on `void *`:

    class HANDLE_converter(CConverter):
        type = 'void *'
        format_unit = '"_Py_PARSE_INTPTR"'

    class HANDLE_return_converter(CReturnConverter):
        type = 'void *'

        def render(self, function, data):
            self.declare(data)
            self.err_occurred_if("_return_value == INVALID_HANDLE_VALUE", data)
            data.return_conversion.append(
                'return_value = PyLong_FromVoidPtr(_return_value);\n')

Then change the `open_osfhandle` parameter to `handle: HANDLE` and the `get_osfhandle` return type to `-> HANDLE`.

segevfiner committed:
bpo-30555: msvcrtmodule.c: Use an argument clinic convertor instead of a cast

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

segevfiner commented:
Merged from master since @haypo fixed the ref leak separately in https://github.com/python/cpython/pull/2003.

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

segevfiner commented:
Merged from master due to argument clinic changes.

segevfiner committed:
bpo-30555: Also consider NULL as a possible error in HANDLE_return_converter

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

zooba commented on Python/fileutils.c:
```diff
@@ -1596,4 +1586,40 @@ _Py_set_blocking(int fd, int blocking)
+{
+    _Py_BEGIN_SUPPRESS_IPH
+    return (void*)_get_osfhandle(fd);
+    _Py_END_SUPPRESS_IPH
```
This can't come after a `return` or code will be unreachable.

zooba commented on Python/fileutils.c:
```diff
@@ -1596,4 +1586,40 @@ _Py_set_blocking(int fd, int blocking)
+{
+    _Py_BEGIN_SUPPRESS_IPH
+    return _open_osfhandle((intptr_t)handle, flags);
+    _Py_END_SUPPRESS_IPH
```
Can't come after a return.

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

segevfiner committed:
Add NEWS.d entry for bpo-30555

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

zooba reviewed (commented):
Hah, just found an unsubmitted review from... a year ago? I guess.

I think they probably still apply, given I don't see any obvious commits below (force pushing after people have reviewed is really annoying, by the way)

segevfiner committed:
Merge remote-tracking branch 'origin/master' into bpo-30555-windowsconsoleio-fd-redirection

segevfiner committed:
Review fix

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

eryksun commented on Modules/_io/winconsoleio.c:
```diff
@@ -466,25 +466,12 @@ _io._WindowsConsoleIO.fileno
-    }
-    if (self->fd < 0)
-        return err_mode("fileno");
     return PyLong_FromLong(self->fd);
```
`error_closed` needs to be called instead of returning `int(-1)`:

```suggestion
    if (self->fd < 0)
        return err_closed();
    return PyLong_FromLong(self->fd);
```

segevfiner committed:
Merge remote-tracking branch 'upstream/master' into bpo-30555-windowsconsoleio-fd-redirection

segevfiner committed:
fileno should error when closed

Co-authored-by: Eryk Sun <eryksun@gmail.com>

segevfiner committed:
Fix faulty call to IOBase.close introduced by automatic merge
